### PR TITLE
Force quality code

### PR DESCRIPTION
**Pull request by sschaef**
This PR fixes all warnings in our codebase and enables a lot of scalac flags, including `-Xfatal-warning`, in our build in order to guarantee that in future no new warnings are introduced into the codebase. Beside from the build, the compiler options are also enabled in the project preferences and therefore directly lead to compilation errors in Eclipse.

The following warnings are enabled:
```
-feature
-unchecked
-deprecation
-language:_
-Xlint
-Xfuture
-Xfatal-warnings
-Yno-adapted-args
-Ywarn-dead-code
-Ywarn-unused-import
-Ywarn-unused
```

I don't know what is the best way to review this PR - probably you need to review every commit separately. Otherwise the changes are too huge. Most changes are cosmetic but I also fixed a lot of deprecation warnings which are more unsafe.

I also introduced the type `org.scalaide.util.internal.Suppress` which allows us to suppress deprecation warnings (what a bad hack, it is a surprise that it is even possible). This is not a very nice solution but some deprecations either couldn't be fixed (Scala Actors) or I had no idea how to do it.

There is only one small downside that comes from enabling `-deprecation`: Whenever the API of scalac changes (or any other API we compile nightly) and a deprecation warning is introduced to a definition we use, our build breaks (and therefore also the `scala/scala` build). But this is only a small downside in my opinion, given that it is easy to alias a definition in the `Suppress` type and then rebuild `scala/scala`.

- [x] Rename `Proxy` to `ProxyType`, not `P`
- [x] Update Eclipse build settings

### CODE REVIEW

**Comment by dragos** [comment]
I would really like to keep `-deprecated`. I think it's important to know this, even though sometimes it's impossible to use the undeprecated method (for instance, when cross-compiling between two Scala version).

By consequence, unfortunately, we can't use `-Xfatal-warnings`.

I also think missing-interpolator is very useful. I don't mind occasional false positives, and let's try to report them to the Scala team.

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/JdiProxy.scala [hunk 1]
```diff
@@ -95,10 +93,10 @@ trait JdiProxy extends Dynamic {
 /**
  * Base for companion objects for [[org.scalaide.debug.internal.expression.proxies.JdiProxy]].
  */
-trait JdiProxyCompanion[Proxy <: JdiProxy, Underlying <: Value] {
+trait JdiProxyCompanion[P <: JdiProxy, Underlying <: Value] {
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/JdiProxy.scala]: Why rename `Proxy` to `P`? I see no warning in this file, and those names were descriptive for a reason.

Comment (sschaef) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/JdiProxy.scala]: `Proxy` shadowed a type with the same name.

Comment (sschaef) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/JdiProxy.scala]: I'm happy to rename them to something more descriptive though.

Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/JdiProxy.scala]: Ok, there is a `scala.Proxy` in every scope, meh. I would either use `import scala.{ Proxy => _ }` or rename it to `ProxyType`.

**Comment by sschaef** [comment]
On 05/14/2015 11:45 AM, Iulian Dragos wrote:
>
> I would really like to keep |-deprecated|. I think it's important to 
> know this, even though sometimes it's impossible to use the 
> undeprecated method (for instance, when cross-compiling between two 
> Scala version).
>
> By consequence, unfortunately, we can't use |-Xfatal-warnings|.
>
I would also like to keep it, but `-Xfatal-warnings` is more important. 
Without it, this PR is more or less useless because we can't ensure that 
no new warnings are introduced in the codebase. Right now, the build 
doesn't even show deprecations, therefore we wouldn't loose anything by 
disabling it. People can and should enable them as warning in the 
compiler preferences in Eclipse.

The thing with deprecated is that we really can't fix it right now:

```
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/completion/ScalaJavaCompletionProposalComputer.scala:88: 
warning: value JLS3 in object AST is deprecated: see corresponding 
Javadoc for more information.
[WARNING]     val ast = unit.reconcile(AST.JLS3,
[WARNING]                                  ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/ScalaAutoIndentStrategy.scala:473: 
warning: value JLS3 in object AST is deprecated: see corresponding 
Javadoc for more information.
[WARNING]       val parser = ASTParser.newParser(AST.JLS3)
[WARNING]                                            ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/compiler/ScalaPresentationCompiler.scala:493: 
warning: overriding method inform in trait Reporting is deprecated: This 
forwards to the corresponding method in reporter -- override reporter 
instead
[WARNING]   override def inform(msg: String): Unit =
[WARNING]                ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:82: 
warning: trait IBufferFactory in package core is deprecated: see 
corresponding Javadoc for more information.
[WARNING]   override def getBufferFactory() : IBufferFactory = 
classFile.getBufferFactory()
[WARNING]                                     ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:82: 
warning: method getBufferFactory in class Openable is deprecated: see 
corresponding Javadoc for more information.
[WARNING]   override def getBufferFactory() : IBufferFactory = 
classFile.getBufferFactory()
[WARNING] ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:129: 
warning: trait ICodeCompletionRequestor in package core is deprecated: 
see corresponding Javadoc for more information.
[WARNING]   def codeComplete(offset : Int, requestor : 
ICodeCompletionRequestor): Unit = { classFile.codeComplete(offset, 
requestor) }
[WARNING]                                              ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:129: 
warning: method codeComplete in class ClassFile is deprecated: see 
corresponding Javadoc for more information.
[WARNING]   def codeComplete(offset : Int, requestor : 
ICodeCompletionRequestor): Unit = { classFile.codeComplete(offset, 
requestor) }
[WARNING] ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:130: 
warning: trait ICompletionRequestor in package core is deprecated: see 
corresponding Javadoc for more information.
[WARNING]   def codeComplete(offset : Int, requestor : 
ICompletionRequestor): Unit = { classFile.codeComplete(offset, requestor) }
[WARNING]                                              ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:130: 
warning: method codeComplete in class ClassFile is deprecated: see 
corresponding Javadoc for more information.
[WARNING]   def codeComplete(offset : Int, requestor : 
ICompletionRequestor): Unit = { classFile.codeComplete(offset, requestor) }
[WARNING] ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:133: 
warning: trait ICompletionRequestor in package core is deprecated: see 
corresponding Javadoc for more information.
[WARNING]   def codeComplete(offset : Int, requestor : 
ICompletionRequestor, owner : WorkingCopyOwner): Unit = { 
classFile.codeComplete(offset, requestor, owner) }
[WARNING]                                              ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:133: 
warning: method codeComplete in class ClassFile is deprecated: see 
corresponding Javadoc for more information.
[WARNING]   def codeComplete(offset : Int, requestor : 
ICompletionRequestor, owner : WorkingCopyOwner): Unit = { 
classFile.codeComplete(offset, requestor, owner) }
[WARNING] ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:145: 
warning: trait IBufferFactory in package core is deprecated: see 
corresponding Javadoc for more information.
[WARNING]   def findSharedWorkingCopy(bufferFactory : IBufferFactory) : 
IJavaElement = { throw new UnsupportedOperationException }
[WARNING]                                             ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:151: 
warning: trait IBufferFactory in package core is deprecated: see 
corresponding Javadoc for more information.
[WARNING]     factory : IBufferFactory,
[WARNING]               ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:156: 
warning: trait IBufferFactory in package core is deprecated: see 
corresponding Javadoc for more information.
[WARNING]     factory : IBufferFactory,
[WARNING]               ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/CompilationUnitAdapter.scala:157: 
warning: method getWorkingCopy in class ClassFile is deprecated: see 
corresponding Javadoc for more information.
[WARNING]     problemRequestor : IProblemRequestor) : IJavaElement = 
classFile.getWorkingCopy(monitor, factory)
[WARNING] ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/logging/StreamRedirect.scala:18: 
warning: method setOut in class DeprecatedConsole is deprecated: Use withOut
[WARNING]       Console.setOut(outStream)
[WARNING]               ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/logging/StreamRedirect.scala:26: 
warning: method setOut in class DeprecatedConsole is deprecated: Use withOut
[WARNING]       Console.setOut(defaultStdOut)
[WARNING]               ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/logging/StreamRedirect.scala:36: 
warning: method setErr in class DeprecatedConsole is deprecated: Use withErr
[WARNING]       Console.setErr(errStream)
[WARNING]               ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/logging/StreamRedirect.scala:44: 
warning: method setErr in class DeprecatedConsole is deprecated: Use withErr
[WARNING]       Console.setErr(defaultStdErr)
[WARNING]               ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ScalaProject.scala:383: 
warning: method findContainersForLocation in trait IWorkspaceRoot is 
deprecated: see corresponding Javadoc for more information.
[WARNING] EclipseUtils.workspaceRoot.findContainersForLocation(path) match {
[WARNING]                                  ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/core/internal/repl/EclipseRepl.scala:174: 
warning: trait Actor in package actors is deprecated: Use the akka.actor 
package instead. For migration from the scala.actors package refer to 
the Actors Migration Guide.
[WARNING] class EclipseRepl(client: Client, builder: Builder) extends Actor
[WARNING] ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/ScalaSourceFileEditor.scala:294: 
warning: trait SynchronizedBuffer in package mutable is deprecated: 
Synchronization via traits is deprecated as it is inherently 
unreliable.  Consider java.util.concurrent.ConcurrentLinkedQueue as an 
alternative.
[WARNING]     private val reconcilingListeners = new 
ArrayBuffer[IJavaReconcilingListener] with 
SynchronizedBuffer[IJavaReconcilingListener]
[WARNING] ^
[ERROR] error: No warnings can be incurred under -Xfatal-warnings.
```

Especially going away from Actors will take us some time.

> I also think missing-interpolator is very useful. I don't mind 
> occasional false positives, and let's try to report them to the Scala 
> team.
>
I'm not sure it these are even bugs in the compiler (because I can't 
reproduce them in a standalone environment) or in the scala-maven-plugin 
(I can't think of how that could happen there). And false positives are 
a no go with `-Xfatal-warnings`.

```
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/ui/internal/preferences/SyntaxColoringPreferencePage.scala:84: 
warning: possible missing interpolator: detected interpolated identifier 
`$SCALADOC_MACRO`
[WARNING]     ("""package foo.bar.baz
[WARNING]      ^
[WARNING] 
/home/antoras/dev/scala/scala-ide/org.scala-ide.sdt.core/src/org/scalaide/ui/internal/wizards/NewFileWizard.scala:241: 
warning: possible missing interpolator: detected an interpolated expression
[WARNING]         val len = v.getName().length() + "${}".length()
[WARNING]                                          ^
```

**Comment by Kwestor** [comment]
I just checked this code myself, rebuilt it and it fails with:

> extension of type scala.Dynamic needs to be enabled by making the implicit value scala.language.dynamics visible. This can be achieved by adding the import clause 'import scala.language.dynamics' or by setting the compiler option -language:dynamics. See the Scala docs for value scala.language.dynamics for a discussion why the feature needs to be explicitly enabled.	ReflectAccess.scala	/org.scala-ide.sdt.core/src/org/scalaide/util/internal	line 29

Should eclipse be configured for this changes by default after rebuild?

**Comment by sschaef** [comment]
The Eclipse build settings for Scala IDE are not yet changed. I'm going to do that.

**Comment by Kwestor** [comment]
Ok. I've added this to PR description.

**Comment by sschaef** [comment]
This PR has changed its fate greatly - it is now a lot more difficult to review than it was before. But were would be the fun if it would be easy?

See the PR message for more information.

**Comment by Kwestor** [comment]
Could you please make a separate PR for only commit 46240695646d68c18f8cdfa0a2b409eb6ebdd74e? One with `: Unit = ` added everywhere. I like co review this PR, but it's technically not possible right now.

**Comment by dragos** [comment]
By far the most common mistake that scalac can flag for me is missing a string interpolator. If we need to rewrite the code (there's only one instance of a false positive in this codebase) to go around it and enable it, we should. Otherwise this `-XfatalWarnings` where you disable the most common  mistake isn't very useful.

**Comment by sschaef** [comment]
@dragos Yes, I already did some code rewritings. The flag for missing interpolators is now also enabled.

**Comment by sschaef** [comment]
@Kwestor I removed the mentioned commit. Superseded by #953

**Comment by Kwestor** [comment]
@sschaef Thanks. Do you plan to make a PR for those Units? I like them, but on separate PR to make it more readable.

**Comment by sschaef** [comment]
Yes, I'm going to open another PR for them.

**Pull request closed without merging**